intrusive_collections_examples/src/basic_datastructures/red_black_tree.rs

```diff
@@ -0,0 +1,359 @@
+/// changed.  For a more in-depth explanation of why this is so, read the
+/// [`README.md`](../README.md) file.
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct RedBlackTreeNode<T>
+where
+    T: Debug + Clone + Default + PartialEq + Eq + PartialOrd + Ord + Hash,
+{
+    /// This is the key used to compare different nodes in the tree.  While it
+    /// is possible to defer this to the object that owns the node, we're going
+    /// to use the simpler route and store the key in here.
+    key: T,
+
+    /// All red-black tree nodes are colored either red or black.
+    ///
+    /// Our chosen default color is black, which means that when we create a
+    /// single root node, it is a complete tree by itself.
+    color: RedBlackTreeNodeColor,
+
+    /// The node that is considered to be the parent of this node.  If there is
+    /// no parent, then this will be `None`.
+    parent: Option<*const RedBlackTreeNode<T>>,
+
+    /// The node that is considered to be the left child of this node.  If there
+    /// is no left child, then this will be `None`.
+    left_child: Option<*const RedBlackTreeNode<T>>,
+
+    /// The node that is considered to be the right child of this node.  If
+    /// there is no right child, then this will be `None`.
+    right_child: Option<*const RedBlackTreeNode<T>>,
+
+    /// Since we're using raw pointers, we **cannot** allow these instances to
+    /// be moved by the compiler (dropping is OK, but moving will almost
+    /// certainly cause UB).  For more information on why this is necessary,
+    /// start reading at
+    /// https://doc.rust-lang.org/std/pin/index.html#example-self-referential-struct
+    /// and be extra careful to read all the way through the explanations about
+    /// drop!
+    _pin: PhantomPinned,
+}
+
+impl<T> Drop for RedBlackTreeNode<T>
+where
+    T: Debug + Clone + Default + PartialEq + Eq + PartialOrd + Ord + Hash,
+{
+    /// Our custom Drop implementation to fix up our neighbors' pointers.
+    ///
+    /// When a node is dropped, it leaves a hole in the tree.  All of the node's
+    /// neighbors (parents and children) will have pointers to a node that no
+    /// longer exists.  We want to fix them up by making them forget about this
+    /// node, if they haven't already forgotten about it.
+    ///
+    /// ## Safety
+    ///
+    /// This is potentially unsafe; there is no guarantee that when we traverse
+    /// the pointers that they are actually pointing to valid memory.  Even if
+    /// they *are* pointing to valid memory, there's no guarantee that we have a
+    /// tree!  If the pointers actually form some kind of arbitrary
+    /// non-tree-like graph, arbitrarily bad things can happen, which is why we
+    /// need to use `unsafe` block below.  In addition, I'm casting away the
+    /// mutability/immutability constraints, which can trick the compiler into
+    /// doing very weird things.
+    fn drop(&mut self) {
```

Comment:
It's difficult to tell if this unsafe code is actually correct, once you have a more complete implementation we'll need to run it through MIRI with some tests to give some basic assurances. Currently, MIRI can't catch aliasing violations if you stick to raw pointers, so it won't be a perfect test, but it should get us started.